internal/discovery/discovery.go

```diff
@@ -0,0 +1,138 @@
+limitations under the License.
+*/
+
+// Package discovery provides a discovery and resolution logic for GVKs.
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+// PopulateTimeout is the timeout on populating the cache for the first time.
+const PopulateTimeout = 10 * time.Second
+
+// Interval is the time interval between two cache sync checks.
+const Interval = 3 * time.Second
+
+// StartDiscovery starts the discovery process, fetching all the objects that can be listed from the apiserver, every `Interval` seconds.
+// resolveGVK needs to be called after StartDiscovery to generate factories.
+func (r *CRDiscoverer) StartDiscovery(ctx context.Context, config *rest.Config) error {
+	dClientset := dynamic.NewForConfigOrDie(config)
+	factory := dynamicinformer.NewFilteredDynamicInformer(dClientset, schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}, "", 0, nil, nil)
+	informer := factory.Informer()
+	stopper := make(chan struct{})
+	_, err := informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			for _, version := range obj.(*unstructured.Unstructured).Object["spec"].(map[string]interface{})["versions"].([]interface{}) {
+				gotGVK := schema.GroupVersionKind{
+					Group:   obj.(*unstructured.Unstructured).Object["spec"].(map[string]interface{})["group"].(string),
+					Version: version.(map[string]interface{})["name"].(string),
+					Kind:    obj.(*unstructured.Unstructured).Object["spec"].(map[string]interface{})["names"].(map[string]interface{})["kind"].(string),
+				}
+				r.AppendToMap(gotGVK)
+				r.SafeWrite(func() {
+					r.WasUpdated = true
+				})
+			}
+		},
+	})
+	if err != nil {
+		return err
+	}
+	// Respect context cancellation.
+	go func() {
+		for range ctx.Done() {
+			klog.Infoln("Context cancelled, stopping discovery")
+			close(stopper)
+			return
+		}
+	}()
+	go informer.Run(stopper)
+	return nil
+}
+
+// ResolveGVK resolves the variable VKs to a GVK list, based on the current cache.
+func (r *CRDiscoverer) ResolveGVK(gvk schema.GroupVersionKind) (resolvedGVKs []schema.GroupVersionKind, err error) {
+	g := gvk.Group
+	v := gvk.Version
+	k := gvk.Kind
+	if g == "" || g == "*" {
+		err = fmt.Errorf("group is required in the defined GVK %v", gvk)
+		return nil, err
+	}
+	hasVersion := v != "" && v != "*"
+	hasKind := k != "" && k != "*"
+	// No need to resolve, return.
+	if hasVersion && hasKind {
+		return []schema.GroupVersionKind{
+			{Group: g, Version: v, Kind: k},
+		}, nil
+	}
+	if hasVersion && !hasKind {
+		kinds := r.Map[g][v]
+		for _, kind := range kinds {
+			if kind == "Scale" || kind == "Status" {
```

Comment:
Can you explain why these get skipped?

Reply:
Had these checks earlier because the discovery package (in client-go) had methods that we were using to discover all GVKs also had these subresources present as kinds in it. Not needed anymore. 👍🏼